### PR TITLE
Feature: Reworked record combiner to support redefines in the writer

### PR DESCRIPTION
**Pull request by Il-Pela**
## Summary

This PR adds `REDEFINES` support to `NestedRecordCombiner`. Previously, the writer silently **dropped** any field declared with `REDEFINES` when building its internal AST (`stmt.redefines.isEmpty` filter), meaning DataFrames could only ever populate the base field of a redefined region — any attempt to write via a redefining field was ignored or failed under strict schema validation.

## What changed

- `WriterAst.scala`
  - Added `RedefineAlternative(fieldName, ast)` and `RedefineGroup(alternatives, actualSize)` AST node types to represent a set of mutually-exclusive fields sharing the same byte region.

- `NestedRecordCombiner.scala`
  - `buildGroupField` now clusters a base field together with its consecutive chain of `REDEFINES` fields into a single `RedefineGroup` node (mirroring the clustering/sizing already done in `cobol-parser`'s `BinaryPropertiesAdder`), instead of dropping the redefining fields.
  - New `buildRedefineGroup`/`buildChildNode` helpers build each alternative (primitive or nested group) leniently, then apply strict-schema validation once at the cluster level — failing only if **none** of the alternatives are present in the DataFrame schema.
  - `writeToBytes` gained a `RedefineGroup` case:
    - Writes the bytes of the single populated alternative (chosen per-row based on which field is non-null).
    - Leaves the shared region as zero-bytes if no alternative is populated (or throws, under strict schema).
    - When more than one alternative is populated on the same row, the behavior is now **configurable** via the new `write_strict_redefines` option (default `false`):
    - **Non-strict (default):** writes the first populated alternative in copybook declaration order (base field before its redefiners), logs a warning naming the ignored fields, and continues the write.
    - **Strict (`write_strict_redefines=true`):** fails fast with a clear `IllegalArgumentException` naming the conflicting fields, as before.
  - New `isPopulated` helper determines whether an AST node (primitive, group, or nested `RedefineGroup`) has a value for a given row.

## Behavior notes / design decisions

- **Conflict policy** is now configurable instead of always fail-fast: by default, writing a row where two alternatives of the same REDEFINES chain are both non-null writes the first populated alternative (in copybook order) and warns, so a single malformed row no longer aborts the whole write job. Opt into the previous fail-fast behavior with `write_strict_redefines=true`.
- **Byte width follows the widest alternative**: the shared region always reserves the largest alternative's size (consistent with how the parser computes offsets), so a 5-byte base field redefined by a 35-byte field still reserves 35 bytes; unused trailing bytes are binary zeroes (`0x00`), not spaces.
- **REDEFINES works on group fields too**: an alternative can itself be a nested group with its own sub-fields, not just a primitive.
- No changes to `cobol-parser`, `RecordCombinerSelector`, or reading/decoding logic — this is writer-only and fully backward compatible with non-REDEFINES copybooks.

## Testing

Added new tests to `FixedLengthEbcdicWriterSuite`:

1. Write using only the base field (regression baseline).
2. Write using only the redefining field.
3. Fail fast when base + redefine are both populated and `write_strict_redefines=true`. 
4. Write the first alternative (base field) when base + redefine are both populated and strict mode is disabled (default).
5. Write the 3rd alternative of a 3-way REDEFINES chain.
6. Fail fast on conflict between non-adjacent alternatives in a 3-way chain, with `write_strict_redefines=true`.
7. Write the first alternative of a 3-way chain when non-adjacent alternatives conflict and strict mode is disabled (default).
8. Mixed rows in a single DataFrame (2 rows via base field, 2 via redefining field) — validates per-row dispatch.
9. Zero-filled bytes when no alternative is present and `strict_schema=false`.
10. Clear error when no alternative is present and `strict_schema=true` (default).
11. Alternatives of different sizes — validates max-size cluster width and zero-byte padding.
12. REDEFINES on nested group fields, including a `REC-TYPE` discriminator column following the common COBOL convention for tagging which alternative a record uses.

Also added a `CobolParametersParserSuite` case verifying `write_strict_redefines` parses into `WriterParameters.strictRedefines`.

## Files changed

- `spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/NestedRecordCombiner.scala`
- `spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/WriterAst.scala`
- `spark-cobol/src/test/scala/za/co/absa/cobrix/spark/cobol/writer/FixedLengthEbcdicWriterSuite.scala`
- `cobol-parser/src/main/scala/za/co/absa/cobrix/cobol/reader/parameters/WriterParameters.scala`
- `cobol-parser/src/main/scala/za/co/absa/cobrix/cobol/reader/parameters/CobolParametersParser.scala`
- `cobol-parser/src/test/scala/za/co/absa/cobrix/cobol/reader/parameters/CobolParametersParserSuite.scala`

# Final Notes
Please let me know what do you think about this PR and if there is the margin of adding this functionality to the library. I'm open to further communication and collaboration and looking forward to read feedbacks from you.

Co-author of this PR: [Andrea Fonti](https://github.com/andrea-rockt)

Thanks again for the immense work you're doing into maintaining this project.

Talk soon,
Francesco

## Release Notes

* **Improvements**
  * Enhanced handling of COBOL REDEFINES alternatives with better validation and error reporting.
  * Improved data serialization logic to correctly process mutually exclusive field alternatives.

* **Tests**
  * Added comprehensive test coverage for REDEFINES group scenarios, including edge cases and conflict detection.

### CODE REVIEW

File: spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/NestedRecordCombiner.scala [hunk 1]
```diff
@@ -604,6 +538,38 @@ object NestedRecordCombiner {
...
+            actualSize
+          case Seq(only) =>
+            writeToBytes(only.ast, row, ar, currentOffset, variableLengthOccurs, writerParameters)
+            actualSize
+          case multiple =>
+            val fieldNames = multiple.map(_.fieldName).mkString("', '")
+            throw new IllegalArgumentException(
+              s"Conflicting REDEFINES fields populated on the same row: '$fieldNames'. " +
+                s"Only one field of a REDEFINES group can have a non-null value at a time."
+            )
```
Comment (yruslan) [inline spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/NestedRecordCombiner.scala]: Throwing exceptions from inside a Spark job is not a usual practice since this can cancel a job that processes GBs of data just on a single data error. Usually, in Spark throwing exception on data is the last resort. 

I'd prefer when multiple alternatives are possible, just use the first one.

No need to fix it yourself, I can fix the logic once the PR is merged. Up to you.

**Review by yruslan — APPROVED** [review]
This is amazing! The solution is very elegant and solves the very important use case. I like it a lot. Have just 1 suggestion to consider.

File: spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/NestedRecordCombiner.scala (hunk 1 above)
Comment (Il-Pela) [reply spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/NestedRecordCombiner.scala]: Yep, I totally agree with you.
I can work on this between today and tomorrow and update the PR by implementing your suggestion (use the first one when there there are multiple alternatives).

Comment (Il-Pela) [reply spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/NestedRecordCombiner.scala]: Ciao @yruslan , I've updated the PR with your suggestion.
I modeled it with a new writer option that 
- by default is non-strict: when multiple alternatives are possible, takes the first one.
- eventually can be made strict: fails when multiple alternatives are found.

I did that because I agreed with you that from an analytics point of view throwing GBs of data for an error on a single row is not optimal. 
However, from an operational point of view, an aware user, could opt for a stricter management of such scenarios to fail the job to avoid unwanted writes during the process that can potentially harm downstream operations.

I've updated the PR text/references accordingly.

Let me know what you think about this.

Comment (yruslan) [reply spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/NestedRecordCombiner.scala]: Perfect, thank you! Merging...

One additional use case came to mind in regards to redefines. In come copybooks we have redefines that provide multiple 'views' on the same field, e.g.:
```cobol
05  ACCOUNT-NUMBER-FULL  9(10).
05  ACCOUNT-NUMBER-DETAIL  REDEFINES ACCOUNT-NUMBER-FULL.
    10 PREFIX        9(4).
    10 NUMBER        9(6).
```

In this case both alternatives would have values, but they are essentially the same. So having the relaxed redefine strictness by default makes perfect sense.

Comment (Il-Pela) [reply spark-cobol/src/main/scala/za/co/absa/cobrix/spark/cobol/writer/NestedRecordCombiner.scala]: Great!
Always a pleasure to contribute!
